src/app/storage/AttributeStorage.h

```diff
@@ -0,0 +1,161 @@
+
+        const ByteSpan mData; // where the data resides
+        Type mType;
+    };
+
+    /// Represents an attribute value that can be read from storage.
+    class Buffer
+    {
+    public:
+        // Separates out what `mData`
+        enum class Type
+        {
+            kPrimitive,           // Integer/bool/float (generally fixed size memory buffer)
+            kMutableByteSpan,     // Raw ByteSpan
+            kStringOneByteLength, // string where length is never larger than 255 characters
+            kStringTwoByteLength, // string where length is never larger than 0xFFFF characters
+            kBytesOneByteLength,  // byte string where length is never larger than 255 characters
+            kBytesTwoByteLength,  // byte string where length is never larger than 0xFFFF characters
+        };
+
+        Buffer(const Buffer & other) = default;
+
+        Buffer(ShortPascalString<char> & data) : Buffer(data.Buffer().data(), data.Buffer().size(), Type::kStringOneByteLength) {}
+        Buffer(LongPascalString<char> & data) : Buffer(data.Buffer().data(), data.Buffer().size(), Type::kStringTwoByteLength) {}
+
+        Buffer(ShortPascalString<uint8_t> & data) : Buffer(data.Buffer().data(), data.Buffer().size(), Type::kBytesOneByteLength) {}
+        Buffer(LongPascalString<uint8_t> & data) : Buffer(data.Buffer().data(), data.Buffer().size(), Type::kBytesTwoByteLength) {}
+
+        template <typename T>
+        static Buffer Primitive(T & value)
+        {
+            static_assert(sizeof(value) <= 8, "Implementations rely these can be mapped as 8 byte or less");
+            return { &value, sizeof(value), Type::kPrimitive };
+        }
+
+        /// initializes the data as "raw". Note that `data()` in this case
+        /// will return a pointer to the underlying `MutableByteSpan`
+        static Buffer Raw(MutableByteSpan & data) { return { &data, 0, Type::kMutableByteSpan }; }
+
+        void * data() { return mData; }
+        size_t size() const { return mBufferSize; }
+        Type type() const { return mType; }
+
+    private:
+        Buffer(void * data, size_t size, Type type) : mData(data), mBufferSize(size), mType(type) {}
+
+        void * mData;
+        size_t mBufferSize; // size of the data in mData
+        Type mType;
+    };
+
+    virtual ~AttributeStorage() = default;
+
+    /// Writes the given value to persistent storage
+    ///
+    /// It is expected that `Read` will read back the same data as read.
+    virtual CHIP_ERROR Write(const ConcreteAttributePath & path, const Value & value) = 0;
+
+    /// Read the given value from persistent storage.
+    ///
+    /// Notable possible errors:
+    ///    CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND - path does not exist in storage
+    ///    CHIP_ERROR_BUFFER_TOO_SMALL - insufficient storage to fetch the data
+    virtual CHIP_ERROR Read(const ConcreteAttributePath & path, Buffer buffer) = 0;
```

Comment:
Should this take the buffer by reference?  Or at least const ref, so it's not forced to be passed by value?

Some documentation about lifetimes (e.g. Buffer is allowed to die as soon as this returns, so be a call-site temporary, and the data will be in the data bits it pointed to?) might be good.